recipes/openblas/all/conanfile.py

```diff
@@ -68,23 +118,39 @@ def configure(self):
         if self.options.shared:
             self.options.rm_safe("fPIC")
 
+        # When cross-compiling, OpenBLAS requires explicitly setting TARGET
+        if cross_building(self, skip_x64_x86=True) and not self.options.target:
+            # Try inferring the target from settings.arch
+            target = conan_arch_to_openblas_target.get(str(self.settings.arch))
+            if target:
+                self.output.warning(f'Setting OpenBLAS TARGET={target} based on settings.arch. This may result in suboptimal performance. Set the "{self.name}/*:target=XXX" option to silence this warning.')
+                self.options.target = target
```

Comment:
question here!

if we only set `self.options.target` when `cross_building`, that means that given a host profile, this recipe will have different package IDs depending on whether it was crossbuilt or not. This is not ideal - one may expect a "producer" such as CI to be able to cross-build, and a consumer on the "native" target system will not find that package because when not-crossbuild this option will default to `None` - unless I'm mistaken.

I wonder what values are mapped when NOT cross building - but if this file https://github.com/OpenMathLib/OpenBLAS/blob/442dec28df3c04b32e6f8630b40080d6a5715843/cmake/prebuild.cmake#L1402 is any indication, I have a suspicion that it is dynamically setting based on the features of the CPU of the machine doing the build? Then we're also in the same situation: for the same package ID, the contents may be different depending on which machine the build ran on, correct?

If this is the case - would it make sense to _always_ give `self.options.target` a value derived from the `arch` setting - regardless of whether or not it is cross-compiled, and point the user to set one that better suits their needs if necessary? I think that solves that particular concern

Reply:
Yes, I agree. The reason why I set `target=None` by default is to keep the existing behavior and not break things for current users. But as you said, this is a lie, since OpenBLAS will dynamically change the options in that case, and this _should_ result in different package IDs.

To complicate matters even further, there doesn't seem to be a way to disable the automatic detection in OpenBLAS, the logic you linked to is always executed `if (NOT CMAKE_CROSSCOMPILING)`, so we'd need to patch that (and hope nothing else breaks) if we want to explicitly set `TARGET` from Conan.

And then there's `DYNAMIC_ARCH`, which includes multiple versions of each function, and dispatches dynamically based on the target it's being executed on. I haven't tried this myself, and it does of course increase binary size and compile times.

Does Conan have any standard way to deal with different microarchitectures? I suspect that any library that uses SIMD or other modern instructions runs into this specific issue, so maybe this should be standardized (e.g. as a setting in the user's profile). E.g. Spack supports this: https://spack.readthedocs.io/en/latest/basic_usage.html#support-for-specific-microarchitectures. An alternative could be to adopt the System V ABI micro-architecture levels with coarser granularity: https://gitlab.com/x86-psABIs/x86-64-ABI

A quick google search only yields some old discussions without solutions like https://github.com/conan-io/conan/issues/847. This one about OpenBLAS's `TARGET` option is definitely relevant, though: https://github.com/conan-io/conan-center-index/pull/13478

Reply:
thanks @tttapa  - indeed those are the issues.

I think the current PR is fine:
- it keeps the current behaviour
- it gives an ability that wasn't there before

while I think it's worth fixing this issues, since doing so may result in changing the default option value - it's probably best to handle that in a follow-up PR